velox/exec/ProbeOperatorState.h

```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <fmt/format.h>
+#include <string>
+
+namespace facebook::velox::exec {
+
+/// Define the internal execution state for hash/nested loop join probe. The
+/// valid state transition is depicted as follows:
+///
+///                           +--------------------------------+
+///                           ^                                |
+///                           |                                V
+///   kWaitForBuild -->  kRunning  -->  kWaitForPeers --> kFinish
+///         ^                                |
+///         |                                v
+///         +--------------------------------+
+///
+enum class ProbeOperatorState {
```

Comment:
The nested loop join shares the same control flow as hash join? We also plan to support disk spilling for it? Thanks!

Reply:
Nested loop join will not be supporting disk spilling in my PR. Other than spilling, it shares similar control flow as hash join. I found it handy to apply these state to nested loop join and extend the meaning of kWaitForPeers: In nested loop join, kWaitForPeers is for output mismatched build side output when doing right/full join. When operators finish processing all the input(matching and output left mismatched date for full join), they enter kWaitForPeers state to wait for other peer probe operators to finish processing inputs. When all the operators get noMoreInput() call and finished processing its current input, the last finished operator will gather build side match state from all peers and emit mismatched build side output.